internal/jbig2/segments/generic-refinement-region.go

```diff
@@ -0,0 +1,1020 @@
+				if err != nil {
+					return err
+				}
+				w4 = int(temp)
+			}
+			byteIndex++
+		} else {
+			w4 <<= 1
+		}
+	}
+	return nil
+}
+
+func (g *GenericRefinementRegion) getPixel(b *bitmap.Bitmap, x, y int) int {
+	if x < 0 || x >= b.Width {
+		return 0
+	}
+
+	if y < 0 || y >= b.Height {
+		return 0
+	}
+
+	if b.GetPixel(x, y) {
+		return 1
+	}
+	return 0
+}
+
+func (g *GenericRefinementRegion) overrideAtTemplate0(context, x, y, result, minorX int) int {
+	if g.grAtOverride[0] {
+		context &= 0xfff7
+
+		if g.GrAtY[0] == 0 && int(g.GrAtX[0]) >= -minorX {
+			context |= (result >> uint(7-(minorX+int(g.GrAtX[0]))) & 0x1) << 3
+		} else {
+			context |= g.getPixel(g.RegionBitmap, x+int(g.GrAtX[0]), y+int(g.GrAtY[0])) << 3
+		}
+	}
+
+	if g.grAtOverride[1] {
+		context &= 0xefff
+		if g.GrAtY[1] == 0 && int(g.GrAtX[1]) >= -minorX {
+			context |= (result >> uint(7-(minorX+int(g.GrAtX[1]))) & 0x1) << 12
+		} else {
+			context |= g.getPixel(g.ReferenceBitmap, x+int(g.GrAtX[1]), y+int(g.GrAtY[1]))
+		}
+	}
+	return context
+}
+
+func (g *GenericRefinementRegion) parseHeader() (err error) {
+	common.Log.Trace("[GENERIC-REF-REGION] parsing Header...")
+	ts := time.Now()
+	defer func() {
+		if err == nil {
+			common.Log.Trace("[GENERIC-REF-REGION] parsing header finishid in: %d ns", time.Since(ts).Nanoseconds())
+		} else {
+			common.Log.Trace("[GENERIC-REF-REGION] parsing header failed: %s", err)
+		}
+	}()
+
+	if err = g.RegionInfo.parseHeader(); err != nil {
+		return
+	}
+
+	// Bit 2-7
+	_, err = g.r.ReadBits(6) // Dirty Read
+	if err != nil {
+		return
+	}
+
+	// Bit 1
+	g.IsTPGROn, err = g.r.ReadBool()
+	if err != nil {
+		return
+	}
+
+	// Bit 0
+	var templateID int
+	templateID, err = g.r.ReadBit()
+	if err != nil {
+		return
+	}
+	g.TemplateID = int8(templateID)
+
+	switch g.TemplateID {
+	case 0:
+		g.Template = g.t0
+		if err = g.readAtPixels(); err != nil {
+			return
+		}
+	case 1:
+		g.Template = g.t1
+	}
+	return nil
+}
+
+func (g *GenericRefinementRegion) readAtPixels() error {
+	g.GrAtX = make([]int8, 2)
+	g.GrAtY = make([]int8, 2)
+
+	// Byte 0
+	temp, err := g.r.ReadByte()
+	if err != nil {
+		return err
+	}
+	g.GrAtX[0] = int8(temp)
+
+	// Byte 1
+	temp, err = g.r.ReadByte()
+	if err != nil {
+		return err
+	}
+	g.GrAtY[0] = int8(temp)
+
+	// Byte 2
+	temp, err = g.r.ReadByte()
+	if err != nil {
+		return err
+	}
+	g.GrAtX[1] = int8(temp)
+
+	// Byte 3
+	temp, err = g.r.ReadByte()
+	if err != nil {
+		return err
+	}
+	g.GrAtY[1] = int8(temp)
+	return nil
+}
+
+// setParameters sets the parameters for the Generic Refinemenet Region.
+func (g *GenericRefinementRegion) setParameters(
+	cx *arithmetic.DecoderStats, arithmDecoder *arithmetic.Decoder,
+	grTemplate int8, regionWidth, regionHeight int,
+	grReference *bitmap.Bitmap, grReferenceDX, grReferenceDY int,
+	isTPGRon bool, grAtX []int8, grAtY []int8,
+) {
+	common.Log.Trace("[GENERIC-REF-REGION] setParameters")
+	defer func() {
```

Comment:
Could just add the log at the bottom of the method as there are no other exit points. No defer function seems to be needed here.